src/cluster_context.rs

```diff
@@ -0,0 +1,141 @@
+                                Some(ingress.clone())
+                            } else {
+                                None
+                            }
+                        } else {
+                            None
+                        }
+                    }
+                })
+                .collect(),
+        )
+    }
+
+    /// Return the list of `Namespace` resources that exist in the
+    /// cluster.
+    pub fn namespaces() -> Result<Vec<Namespace>> {
+        // TODO (ereslibre): use macros to remove duplication and then
+        // generalize
+        guest::host_call("kubernetes", "namespaces", "list", &Vec::new())
+            .map_err(|err| anyhow!("failed to call namespaces binding: {}", err))
+            .and_then(|namespaces| {
+                Ok(
+                    serde_json::from_str::<List<Namespace>>(std::str::from_utf8(&namespaces)?)
+                        .map_err(|err| anyhow!("failed to unmarshal namespace list: {}", err))?
+                        .items,
+                )
+            })
+    }
+
+    /// Return the list of `Service` resources that exist in the
+    /// cluster.
+    pub fn services(namespace: NamespaceFilter) -> Result<Vec<Service>> {
+        // TODO (ereslibre): use macros to remove duplication and then
+        // generalize
+        Ok(
+            guest::host_call("kubernetes", "services", "list", &Vec::new())
+                .map_err(|err| anyhow!("failed to call services binding: {}", err))
+                .and_then(|services| {
+                    Ok(
+                        serde_json::from_str::<List<Service>>(std::str::from_utf8(&services)?)
+                            .map_err(|err| anyhow!("failed to unmarshal service list: {}", err))?
+                            .items,
+                    )
+                })?
+                .iter()
+                .filter_map(|service| match &namespace {
+                    NamespaceFilter::AllNamespaces => Some(service.clone()),
+                    NamespaceFilter::Namespace(namespace_filter) => {
+                        if let Some(service_namespace) = &service.metadata.namespace {
+                            if namespace_filter == service_namespace {
+                                Some(service.clone())
+                            } else {
+                                None
+                            }
+                        } else {
+                            None
+                        }
+                    }
+                })
+                .collect(),
+        )
+    }
+}
+
+impl ClusterContext {
+    /// Return a specific ingress object with a given name and a
+    /// namespace filter. If the namespace filter allows for more than
+    /// one namespace, the ingress resource found that matches one of
+    /// the namespaces and the given name will be returned.
+    pub fn ingress(namespace: NamespaceFilter, name: &str) -> Result<Option<Ingress>> {
```

Comment:
I have to admit I don't love this signature. I don't know how common, but you could call this with `AllNamespaces` and a name that is duplicated on N namespaces, returning only the first occurence. OTOH, for that corner case, returning `Result<Option<Vec<Ingress>>` or `Result<Vec<Ingress>>` seems too much. @flavio, do you have any suggestion? [1]

Reply:
I would change it to be `Result<Vec<Ingress>>`. When no matches are found the `Vec` would be empty. I don't see the need to wrap the `Vec` into an `Option`, it just makes everything more complicated.

Reply:
I would use this approach also elsewhere [1]

Reply:
I'll adapt the PR,thanks!